redbot/cogs/streams/streamtypes.py

```diff
@@ -280,22 +282,67 @@ def __init__(self, **kwargs):
         self.id = kwargs.pop("id", None)
         self._client_id = kwargs.pop("token", None)
         self._bearer = kwargs.pop("bearer", None)
+        self._rate_limit_resets: set = set()
+        self._rate_limit_remaining: int = 0
         super().__init__(**kwargs)
 
+    async def wait_for_rate_limit_reset(self) -> None:
+        """Check rate limits in response header and ensure we're following them.
+
+        From python-twitch-client and adaptated to asyncio from Trusty-cogs:
+        https://github.com/tsifrer/python-twitch-client/blob/master/twitch/helix/base.py
+        https://github.com/TrustyJAID/Trusty-cogs/blob/master/twitch/twitch_api.py
+        """
+        if self._rate_limit_remaining == 0:
+            current_time = int(time.time())
+            self._rate_limit_resets = set(x for x in self._rate_limit_resets if x > current_time)
+
+            if len(self._rate_limit_resets) > 0:
+                reset_time = list(self._rate_limit_resets)[0]
+                # Calculate wait time and add 0.1s to the wait time to allow Twitch to reset
+                # their counter
+                wait_time = reset_time - current_time + 0.1
+                await asyncio.sleep(wait_time)
+
+    async def get_data(self, url: str, params: dict = {}) -> Tuple[int, dict]:
+        header = {"Client-ID": str(self._client_id)}
+        if self._bearer is not None:
+            header["Authorization"] = f"Bearer {self._bearer}"
+        await self.wait_for_rate_limit_reset()
+        async with aiohttp.ClientSession() as session:
+            try:
+                async with session.get(
+                    url, headers=header, params=params, timeout=aiohttp.ClientTimeout(total=None)
+                ) as resp:
+                    remaining = resp.headers.get("Ratelimit-Remaining")
+                    if remaining:
+                        self._rate_limit_remaining = int(remaining)
+                    reset = resp.headers.get("Ratelimit-Reset")
+                    if reset:
+                        self._rate_limit_resets.add(int(reset))
+
+                    if resp.status == 429:
+                        log.info(
+                            "Ratelimited. Trying again at %s.", datetime.fromtimestamp(int(reset))
+                        )
+                        return await self.get_data(url)
+
+                    if resp.status != 200:
+                        return None, {}
```

Comment:
We still need to return status code here so that the handling of error codes works.
```suggestion
                        return resp.status, {}
```